### DISCUSSION TITLE
album_count in the artist table - Glitchy or Just Me?

### DISCUSSION THREAD

**Discussion by Andrew-McGee** [ampache/ampache#2923] [General]
Hi all,

Can't remember if I raised this previously but I still seem to get some weirdness with the album_count. I pull this from the artist API call and that seems fine but I notice artists who definitely have an album have this set to 0 in the db (artist table).

Shouldn't this get updated when a catalog add is performed and a new album is added? What is the criteria to update it?

As an aside I am also interested to display how many total albums an artist appears on (not just albums they release but compilations too). Any easy way to get this? The hard way is trawling through artist_songs method results and parsing for unique album names.

**Comment by lachlan-00** [comment]
i'm putting in some new queries for catalog updates, let me know how many rows get updated by these.

```
## album_count
UPDATE `artist`, (SELECT COUNT(DISTINCT `album`.`id`) AS `album_count`, `album_artist` FROM `album` LEFT JOIN `catalog` ON `catalog`.`id` = `album`.`catalog` WHERE `catalog`.`enabled` = '1' GROUP BY `album_artist`) AS `album` SET `artist`.`album_count` = `album`.`album_count` WHERE `artist`.`album_count` != `album`.`album_count` AND `artist`.`id` = `album`.`album_artist`;

## song_count
UPDATE `artist`, (SELECT COUNT(`song`.`id`) AS `song_count`, `artist` FROM `song` LEFT JOIN `catalog` ON `catalog`.`id` = `song`.`catalog` WHERE `catalog`.`enabled` = '1' GROUP BY `artist`) AS `song` SET `artist`.`song_count` = `song`.`song_count` WHERE `artist`.`song_count` != `song`.`song_count` AND `artist`.`id` = `song`.`artist`;

## album_group_count
UPDATE `artist`, (SELECT COUNT(DISTINCT CONCAT(COALESCE(`album`.`prefix`, ''), `album`.`name`, COALESCE(`album`.`album_artist`, ''), COALESCE(`album`.`mbid`, ''), COALESCE(`album`.`year`, ''))) AS `album_group_count`, `album_artist` FROM `album` LEFT JOIN `catalog` ON `catalog`.`id` = `album`.`catalog` WHERE `catalog`.`enabled` = '1' GROUP BY `album_artist`) AS `album` SET `artist`.`album_group_count` = `album`.`album_group_count` WHERE `artist`.`album_group_count` != `album`.`album_group_count` AND `artist`.`id` = `album`.`album_artist`;


```

**Comment by mitchray** [comment]
For artist appearances I subtract the artist ```albumcount``` from how many albums are returned from ```action=artist_albums```

**Reply by lachlan-00** [reply]
i've actually come up across **A LOT** of issues with how counts are managed/migrated/etc/omg it's crazy!

i'm working in this pull https://github.com/ampache/ampache/pull/2926 but since i turned off caching i've been coming across all sorts of issues that i never noticed so it's been going through a lot of work this week.

Right now i've fixed up all the count issues but this pull is going to be quite large in all the changes/fixes as this is my first 5 only change so i'm going a bit ham on the database changes.

**Reply by Andrew-McGee** [reply]
Thanks mate - look forward to this when it goes in.

**Reply by lachlan-00** [reply]
merged now. the database updates will take a while  in one go but the counts/times are updated with changes after that instead of on loads

**Reply by Andrew-McGee** [reply]
Nice one - thanks a lot.
